internal/controllers/machine/machine_controller.go

```diff
@@ -338,20 +338,25 @@ func (r *Reconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Clu
 			conditions.MarkTrue(m, clusterv1.DrainingSucceededCondition)
 			r.recorder.Eventf(m, corev1.EventTypeNormal, "SuccessfulDrainNode", "success draining Machine's node %q", m.Status.NodeRef.Name)
 
-			// After node draining, make sure volumes are detached before deleting the Node.
-			if conditions.Get(m, clusterv1.VolumeDetachSucceededCondition) == nil {
-				conditions.MarkFalse(m, clusterv1.VolumeDetachSucceededCondition, clusterv1.WaitingForVolumeDetachReason, clusterv1.ConditionSeverityInfo, "Waiting for node volumes to be detached")
+			// After node draining, make sure either volumes are detached or left alone after volume detach timeout
+			// exceeded before deleting the Node.
+			if conditions.Get(m, clusterv1.VolumeDetachFinishedCondition) == nil {
+				conditions.MarkFalse(m, clusterv1.VolumeDetachFinishedCondition, clusterv1.WaitingForVolumeDetachReason, clusterv1.ConditionSeverityInfo, "Waiting for node volumes to be detached")
 			}
-			if ok, err := r.shouldWaitForNodeVolumes(ctx, cluster, m.Status.NodeRef.Name, m.Name); ok || err != nil {
+			if r.nodeDrainTimeoutExceeded(m) || r.volumeDetachTimeoutExceeded(m) {
```

Comment:
Q: if we have nodeDrainTimeout 10m and volumeDetach=0 (infinit), I would expect the system to wait indefinitely, is that right? if yes, I think that this test is not correct because if I'm not wrong this will stop re-trying after 10m minutes from last transition of the DrainingSucceededCondition.

Reply:
After looking back at this I think that breaking the two operations under different ifs , each one with its own timeout/annotation is cleaner and easier to explain. Ultimately we are fixing a bug
@vincepri opinions?

Reply:
> f we have nodeDrainTimeout 10m and volumeDetach=0 (infinit), I would expect the system to wait indefinitely, is that right? if yes, I think that this test is not correct because if I'm not wrong this will stop re-trying after 10m minutes from last transition of the DrainingSucceededCondition.

yes, giving another thought, nodeDrainTimeoutExceeded condition is not needed IMO and made some more changes to make it bit easier to read and understand, hope it makes sense, really appreciate your opinion.